qiskit/transpiler/passes/routing/basic_swap.py

```diff
@@ -98,12 +108,79 @@ def run(self, dag):
                     # update current_layout
                     for swap in range(len(path) - 2):
                         current_layout.swap(path[swap], path[swap + 1])
-
-            order = current_layout.reorder_bits(new_dag.qubits)
-            new_dag.compose(subdag, qubits=order)
+            # handle control flow operations
+            for node in subdag.control_flow_ops():
+                updated_ctrl_op, cf_layout = self._transpile_controlflow_op(node.op, current_layout)
+                node.op = updated_ctrl_op
+                cf_layer = True
+            if cf_layer:
+                new_dag.compose(subdag)
+                current_layout = cf_layout
+            else:
+                order = current_layout.reorder_bits(new_dag.qubits)
+                new_dag.compose(subdag, qubits=order)
+        self.property_set["final_layout"] = current_layout
 
         return new_dag
 
+    def _transpile_controlflow_op(self, cf_op, current_layout):
+        """default function"""
+        if isinstance(cf_op, IfElseOp):
+            return self._transpile_controlflow_multiblock(cf_op, current_layout)
+        elif isinstance(cf_op, (ForLoopOp, WhileLoopOp)):
+            return self._transpile_controlflow_looping(cf_op, current_layout)
+        return cf_op, current_layout
+
+    def _transpile_controlflow_multiblock(self, cf_op, current_layout):
+        # pylint: disable=cyclic-import
+        from qiskit.transpiler.passes.routing.layout_transformation import LayoutTransformation
+        from qiskit.converters import dag_to_circuit, circuit_to_dag
+
+        block_circuits = []  # control flow circuit blocks
+        block_dags = []  # control flow dag blocks
+        block_layouts = []  # control flow layouts
+
+        for i, block in enumerate(cf_op.blocks):
+            dag_block = circuit_to_dag(block)
+            _pass = BasicSwap(self.coupling_map, initial_layout=current_layout)
+            updated_dag_block = _pass.run(dag_block)
+            block_dags.append(updated_dag_block)
+            block_layouts.append(_pass.property_set["final_layout"].copy())
+        changed_layouts = [current_layout != layout for layout in block_layouts]
+        if not any(changed_layouts):
+            return cf_op, current_layout
```

Comment:
Is it possible to have not changed the layout, but still have had `BasicSwap` insert swaps?

Reply:
I don't believe so.

Reply:
Let me see if I can draw up an example.

```python
qc = qk.QuantumCircuit(3)
cm = qk.transpiler.CouplingMap.from_line(3)

from qiskit.transpiler.passes import BasicSwap

qc.cx(0,1)
qc.cx(0,2)
qc.cx(1,2)

BasicSwap(cm)(qc).draw()
```

```
                          
q_0: ──■───X───────X──────
     ┌─┴─┐ │       │      
q_1: ┤ X ├─X───■───X───■──
     └───┘   ┌─┴─┐   ┌─┴─┐
q_2: ────────┤ X ├───┤ X ├
             └───┘   └───┘
```

The input and output layouts are the same in this case, even though the pass inserted swaps, so we'd have to make sure to return the updated block in this case.